multicast/multicast.go

```diff
@@ -0,0 +1,354 @@
+	close(errCh)
+
+	var err_ret error
+	for fetchdata := range errCh {
+		if fetchdata != nil {
+			err_ret = errors.Join(err_ret, fetchdata)
+		}
+	}
+	return err_ret
+}
+
+// ListSubscriber lists multicast subscribers in every node for the specified multicast group or all multicast groups
+func (m *Multicast) ListSubscriber() error {
+	if m.params.MulticastGroupIP == "" && !m.params.All {
+		return fmt.Errorf("group-ip or all flag must be specified")
+	} else if m.params.MulticastGroupIP != "" && m.params.All {
+		return fmt.Errorf("only one of group-ip or all flag must be specified")
+	}
+
+	var target string
+	if m.params.All {
+		target = "all"
+	} else {
+		target = m.params.MulticastGroupIP
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), m.params.WaitDuration)
+	defer cancel()
+
+	ciliumPods, err := m.listCiliumPods(ctx)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(ciliumPods))
+	wg.Add(len(ciliumPods))
+
+	for _, ciliumPod := range ciliumPods {
+		go func(pod corev1.Pod) {
+			defer wg.Done()
+			// List multicast subscribers
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "subscriber", "list", target}
+			output, err := m.client.ExecInPod(ctx, pod.Namespace, pod.Name, defaults.AgentContainerName, cmd)
+			//ToDo:detect error type and handle properly
+			if err != nil {
+				errCh <- err
+				return
+			}
+			output_string := "Node: " + pod.Spec.NodeName + "\n" + output.String()
+			fmt.Fprintln(m.params.Writer, output_string)
+		}(ciliumPod)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	var err_ret error
+	for fetchdata := range errCh {
+		if fetchdata != nil {
+			err_ret = errors.Join(err_ret, fetchdata)
+		}
+	}
+	return err_ret
+}
+
+// AddAllNodes add CiliumInternalIPs of all nodes to the specified multicast group as subscribers in every cilium-agent
+func (m *Multicast) AddAllNodes() error {
+	if m.params.MulticastGroupIP == "" {
+		return fmt.Errorf("group-ip must be specified")
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), m.params.WaitDuration)
+	defer cancel()
+	ciliumPods, err := m.listCiliumPods(ctx)
+	if err != nil {
+		return err
+	}
+
+	//Create a map of ciliumInternalIPs of all nodes
+	ipToPodMap := make(map[v2.NodeAddress]string)
+	ipToNodeMap := make(map[v2.NodeAddress]string)
+
+	var wg1 sync.WaitGroup
+	errCh1 := make(chan error, len(ciliumPods))
+	wg1.Add(len(ciliumPods))
+
+	for _, ciliumPod := range ciliumPods {
+		go func(pod corev1.Pod) {
+			defer wg1.Done()
+			ciliumInternalIP, err := m.getCiliumInternalIP(pod.Spec.NodeName)
+			if err != nil {
+				errCh1 <- err
+				return
+			}
+			ipToPodMap[ciliumInternalIP] = pod.Name
+			ipToNodeMap[ciliumInternalIP] = pod.Spec.NodeName
+		}(ciliumPod)
+	}
+
+	wg1.Wait()
+	close(errCh1)
+
+	var err_ret error
+	for fetchdata := range errCh1 {
+		if fetchdata != nil {
+			err_ret = errors.Join(err_ret, fetchdata)
+		}
+	}
+	if err_ret != nil {
+		return err_ret
+	}
+
+	var wg2 sync.WaitGroup
```

Comment:
I'd recommend splitting this into a new function, if not purely for the sake of readability and avoiding the pitfalls of `wg1, wg2..`